move secure boot setup to imghelper

jpculp opened the pull request:
**Description of changes:**

Moves code-blocks related to secure boot to `imghelper`, with some minor adjustments for portability.

**Testing done:**

- Built and smoke tested `aws-k8s-1.28`

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

jpculp committed:
rpm2img: move signing prerequisites to imghelper

Signed-off-by: Patrick J.P. Culp <jpculp@amazon.com>

jpculp commented on twoliter/embedded/imghelper:
```diff
@@ -108,6 +110,214 @@ mkfs_data_xfs() {
+  grub_cfg_sig="${grub_cfg}.sig"
+  [[ -f "${grub_cfg_sig}" ]] && rm "${grub_cfg_sig}"
+  gpg --batch --no-tty --detach-sign "${grub_cfg}"
+  gpg --batch --no-tty --verify "${grub_cfg_sig}" "${grub_cfg}"
```
The reason for the change in syntax here is when `--batch` is not used `gpg` can make assumptions about the matching data files from a single argument. This is discouraged and instead we should always specify both explicitly. Since we're not interfacing with `gpg` interactively anyway, I just went ahead and set all the `gpg` commands to `--batch --no-tty`, which results in more useful errors when something goes wrong.

jpculp commented on twoliter/embedded/imghelper:
```diff
@@ -108,6 +110,214 @@ mkfs_data_xfs() {
+    --remove-section ".pubkey"
+}
+
+uefi_copycert() {
+  local efi_image
+  efi_image="${1:?}"
+  mcopy -i "${efi_image}" "${SBKEYS}"/db.{crt,cer} ::/EFI/BOOT
+}
```
This feels a tiny bit silly, but by moving it out of `rpm2img` we no longer need **SBKEYS** outside of `imghelper`.

jpculp commented on twoliter/embedded/rpm2img:
```diff
@@ -251,126 +249,19 @@ mv "${ROOT_MOUNT}/boot/efi"/* "${EFI_MOUNT}"
-  gpg --list-keys
 fi
 
 pushd "${EFI_MOUNT}/EFI/BOOT" >/dev/null
```
```suggestion

```
I'll move this into the functions.

jpculp commented:
- Removed assumption that the caller will change directory to `${EFI_MOUNT}/EFI/BOOT` before calling UEFI helpers.
- Removed circular references.

rpkelly commented on twoliter/embedded/imghelper:
```diff
@@ -108,6 +110,228 @@ mkfs_data_xfs() {
+  # done when the certificates are created, but the resulting binary files are
+  # not as nice to store in source control.
+  for cert in PK KEK db vendor; do
+    openssl x509 \
+      -inform PEM -in "${SBKEYS}/${cert}.crt" \
+      -outform DER -out "${SBKEYS}/${cert}.cer"
```
Should we be validating that this output is equal to the input somehow, or is it fine to just trust this conversion happened correctly?

jpculp replied:
I think we can trust this conversion (at least as much as we can trust any other conversion in `rpm2img`/`img2img`). We can do a string comparison of some of the contents using `openssl x509` to print, but I'm not sure if we'd want that potentially logged.

rpkelly commented on twoliter/embedded/imghelper:
```diff
@@ -108,6 +110,228 @@ mkfs_data_xfs() {
+
+  # Set AWS environment variables from build secrets, if present.
+  local var val
+  for var in AWS_ACCESS_KEY_ID AWS_SECRET_ACCESS_KEY AWS_SESSION_TOKEN; do
```
Does `AWS_DEFAULT_REGION` also need to be in this list?

jpculp replied:
The region isn't currently a build secret passed to the container. I suspect it's not necessary to add, but if so we'd want to address it in a separate PR.

rpkelly commented on twoliter/embedded/imghelper:
```diff
@@ -108,6 +110,228 @@ mkfs_data_xfs() {
+  shim_sign_key=(-c shim-sign-key)
+  code_sign_key=(-c code-sign-key)
+
+  sbsetup_wrapup local
```
I spent too long trying to figure out how `local` was working here. Maybe just a me thing, but, a comment would be nice :)

jpculp replied:
I'll post a revision and bring back the named variable for readability.

rpkelly commented on twoliter/embedded/imghelper:
```diff
@@ -108,6 +110,228 @@ mkfs_data_xfs() {
+  cert_table "${SBKEYS}/vendor.cer" "${SBKEYS}/vendor.obj"
+
+  # Replace the embedded vendor certificate, then sign shim with the db key.
+  if [[ "$(pesign -i "${shim}" -l)" == "No signatures found." ]]; then
+    mv "${shim}" "${shim}.unsigned"
+  else
+    pesign -i "${shim}" -o "${shim}.unsigned" -u 0 -r
+  fi
+  "${arch}-objdumpcopy" "${shim}.unsigned" \
+    --update-section ".vendor_cert=${SBKEYS}/vendor.obj"
+  pesign -i "${shim}.unsigned" -o "${shim}.signed" -s "${shim_sign_key[@]}"
+  mv "${shim}.signed" "${shim}"
+  rm "${shim}.unsigned"
+  pesign -i "${shim}" -l
+  pesigcheck -i "${shim}" -n 0 -c "${SBKEYS}/db.cer"
```
This block of code is repeated at least 3 times (possibly a 4th in the `vmlinuz_sbsetup` function). Can it be turned into its own function?

jpculp replied:
They are close, but different enough where it wouldn't be as clean as you'd think. It could be split into a `remove_pesig` and `add_pesign` but might end up being about the same number of lines the end. I don't mind adding it though.

bcressey commented on twoliter/embedded/imghelper:
```diff
@@ -108,6 +110,228 @@ mkfs_data_xfs() {
+  mcopy -i "${efi_image}" "${SBKEYS}"/db.{crt,cer} ::/EFI/BOOT
+}
+
+vmlinuz_sbsetup() {
```
Please name these functions more consistently - `vmlinuz` is also an EFI binary and also signed with `pesign`, so it's strange to have `uefi_pesign` and `vmlinuz_sbsetup`.

You should also pick either `noun_verb` or `verb_noun` and stick with it. `verb_noun` is more readable and consistent with other functions already in this file.

bcressey commented on twoliter/embedded/imghelper:
```diff
@@ -108,6 +110,228 @@ mkfs_data_xfs() {
+      >"${vmlinuz%/*}.${vmlinuz##*/}.hmac"
+}
+
+sign_grub() {
```
The function is named `sign_grub` but you're not actually signing grub but rather `grub.cfg` ...

bcressey commented on twoliter/embedded/rpm2img:
```diff
@@ -251,128 +249,20 @@ mv "${ROOT_MOUNT}/boot/efi"/* "${EFI_MOUNT}"
-  pesign -i "${grub}" -o "${grub}.signed" -s "${CODE_SIGN_KEY[@]}"
-  mv "${grub}.signed" "${grub}"
-  pesigcheck -i "${grub}" -n 0 -c "${SBKEYS}/vendor.cer"
+  uefi_pesign "${EFI_MOUNT}" "${ARCH}" SHIM_SIGN_KEY CODE_SIGN_KEY
```
You should refactor this along the lines of `<sign_function> <ARTIFACT>` for consistency with the other signing functions.

You are also not very consistent about when you pass in keys (here) versus when you have the function decide what key to use (`sign_grub`). I would recommend just specializing all of the functions to sign a particular artifact and have the function pick the key to use. There is only one correct choice for a given artifact - using the wrong key will either result in an unbootable image or weaken security.

bcressey commented on twoliter/embedded/imghelper:
```diff
@@ -108,6 +110,228 @@ mkfs_data_xfs() {
+  else
+    pesign -i "${grub}" -o "${grub}.unsigned" -u 0 -r
+  fi
+  "${arch}-objdumpcopy" "${grub}.unsigned" \
+    --file-alignment 4096 \
+    --update-section ".pubkey=${SBKEYS}/config-sign.pubkey"
+  pesign -i "${grub}.unsigned" -o "${grub}.signed" -s "${code_sign_key[@]}"
+  mv "${grub}.signed" "${grub}"
+  rm "${grub}.unsigned"
+  pesign -i "${grub}" -l
+  pesigcheck -i "${grub}" -n 0 -c "${SBKEYS}/vendor.cer"
```
As @rpkelly mentioned, the duplication here is a sign that this function is doing too much, which is part of why its signature is inconsistent.

jpculp commented:
- Split `uefi_pesign` into several more targeted functions.
- Renamed some things for consistency and clarity.

bcressey commented on twoliter/embedded/rpm2img:
```diff
@@ -251,81 +249,13 @@ mv "${ROOT_MOUNT}/boot/efi"/* "${EFI_MOUNT}"
+  # shellcheck disable=SC2034  # variables are used externally by reference
+  declare -a SHIM_SIGN_KEY CODE_SIGN_KEY
+  if [[ -s "${HOME}/.config/aws-kms-pkcs11/config.json" ]]; then
+    sbsetup_aws_profile SHIM_SIGN_KEY CODE_SIGN_KEY
```
Do we still need `SHIM_SIGN_KEY` and `CODE_SIGN_KEY` to be passed in by reference, or can they be set and saved in `imghelper`?

As a minor nit: we could also move this branch into a single `sbsetup_signing_profile` function that calls the right function (aws or local).

jpculp commented:
Rebased.

jpculp committed:
rpm2img: move UEFI signing to imghelper

Signed-off-by: Patrick J.P. Culp <jpculp@amazon.com>

jpculp committed:
rpm2img: move boot signing to imghelper

Signed-off-by: Patrick J.P. Culp <jpculp@amazon.com>

jpculp commented:
- Moved **SHIM_SIGN_KEY** and **CODE_SIGN_KEY** declaration to `imghelper`.
- Wrapped profile setup under a single `sbsetup_signing_profile` function.
- Moved setup a few lines down to share the **UEFI_SECURE_BOOT** conditional.
